### PR TITLE
Fixed restricted editing mode integration with drag and drop.

### LINKED ISSUE #9279: Drag & drop: it's possible to drag content outside restricted editing area
## 📝 Provide detailed reproduction steps (if any)

1. Go to http://localhost:8080/ckeditor5/latest/features/restricted-editing.html
2. Switch to restricted mode
3. Select some content in restricted area and drag it outside

### ✔️ Expected result

Content should not be dragged outside restricted area

### ❌ Actual result

Content is dragged outside restricted area:  
  
0_tp3

## 📃 Other details

* Browser: any

---

If you'd like to see this fixed sooner, add a 👍 reaction to this post.

### LINKED ISSUE #9286: Drag & drop: dragging restricted region into another creates a duplicate
## 📝 Provide detailed reproduction steps (if any)

1. Go to http://localhost:8080/ckeditor5/latest/features/restricted-editing.html
2. Switch to restricted editing mode
3. Fully select some restricted editing area
4. Drag it into another restricted area

### ✔️ Expected result

Only text is dropped, like when doing cut/paste

### ❌ Actual result

A duplicated restricted area is created inside:  
  
0_tp4

## 📃 Other details

* Browser: any

---

If you'd like to see this fixed sooner, add a 👍 reaction to this post.

### PR DESCRIPTION
**Pull request by niegowski**
### Suggested merge commit message ([convention](https://ckeditor.com/docs/ckeditor5/latest/framework/guides/contributing/git-commit-message-convention.html))

Internal (restricted-editing): Fixed restricted editing mode integration with drag and drop. Closes #9286. Closes #9279.

---

### Additional information

### CODE REVIEW

**Comment by niegowski** [comment]
@ckeditor/qa-team could you check if it solves the problem?

**Comment by FilipTokarski** [comment]
In Safari I have disappearing content on drop, but I guess this is some browser thing, right?

https://user-images.githubusercontent.com/34380544/111749291-da5b8b00-8891-11eb-850c-b6f27729aa68.mp4

**Comment by FilipTokarski** [comment]
Apart from that, everything seems to work fine now 👌

**Comment by niegowski** [comment]
> In Safari I have disappearing content on drop, but I guess this is some browser thing, right?

Unfortunately, it's the case that Safari is correctly triggering the `dragend` event. I need to look at it.

File: packages/ckeditor5-restricted-editing/src/restrictededitingmode/converters.js [hunk 1]
```diff
@@ -78,10 +78,10 @@ export function resurrectCollapsedMarkerPostFixer( editor ) {
 	return writer => {
 		let changeApplied = false;
 
-		for ( const [ name, data ] of editor.model.document.differ._changedMarkers ) {
+		for ( const { name, data } of editor.model.document.differ.getChangedMarkers() ) {
 			if ( name.startsWith( 'restrictedEditingException' ) && data.newRange.root.rootName == '$graveyard' ) {
 				writer.updateMarker( name, {
-					range: writer.createRange( writer.createPositionAt( editor.model.document.selection.focus ) )
+					range: writer.createRange( writer.createPositionAt( data.oldRange.start ) )
```
Comment (Reinmar) [inline packages/ckeditor5-restricted-editing/src/restrictededitingmode/converters.js:84]: `data.oldRange` may not be up to date any more if the drop position is before it (same paragraph).

**Comment by Reinmar** [comment]
Two followups:

*   The Safari case.
*   The `data.oldRange` case.

**Pull request merged**